### DISCUSSION TITLE
MSI or MSI & EXE?

### DISCUSSION THREAD

**Discussion by russellbanks** [microsoft/winget-pkgs#68595] [Q&A] [answered]
If a program has two installers, an MSI and an EXE, that both install and essentially do the exact same thing, should only the MSI be provided or should both be provided? Is it good practice to provide both in case a user has a preference or just the MSI because you can guarantee that it's an installer?

**Comment by Trenly** [comment, marked as answer]
Best practice is to provide both, as winget will try to match the install technology if the user has a previously installed version of the package on their machine. There is even a specific error for when a user used the MSI and we only have the exe installer  (And vice versa?) “The install technology differs from the one originally used to install the package. Please uninstall and re-install” (Or something like that, I forget the exact wording)

**Reply by mdanish-kh** [reply]
> There is even a specific error for when a user used the MSI and we only have the exe installer (And vice versa?) 

Vice versa too. https://github.com/microsoft/winget-pkgs/issues/61777

**Reply by russellbanks** [reply]
Just another question - should this be taken the same if a program has an exe installer and a portable exe (because they would have had to have install it with winget to get the ARP entries for the portable)?

**Reply by Trenly** [reply]
> Just another question - should this be taken the same if a program has an exe installer and a portable exe (because they would have had to have install it with winget to get the ARP entries for the portable)?

Yes
